### PR TITLE
gh-1838 added checks for pipe streams

### LINKED ISSUE #1838: Return a popen stream in the response
Hello,

I am trying to return a popen stream like this:

``` php
$stream = \GuzzleHttp\Psr7\stream_for(
    popen(
        'ls', 'r'
    )
);
return $response->withBody($stream);
```

But I see two problems with it:
- For some reason, App::finalize() seems to replace $response->getBody()->stream with an _Unknown resource_.
- `$response->getBody()->getSize()` always returns 0 in this case so App::respond() does not output anything.

### PR DESCRIPTION
**Pull request by llvdl**
Fixes #1838

### CODE REVIEW

**Comment by ghost** [comment]
Since Slim's implementation has methods beyond what's in StreamInterface anyway, I'd add isSeekableForward() and seekForward() for completeness. If one's dealing with pipes, he/she might need them anyway.

```
    /**
     * Returns whether or not the stream is seekable forward.
     *
     * Note: This method is not part of the PSR-7 standard
     * and is provided for Stream's compatibility with pipes.
     *
     * @return bool
     */
    public function isSeekableForward()
    {
        return $this->isSeekable() || $this->isPipe();
    }

    /**
     * Seek forward in the stream.
     *
     * Note: This method is not part of the PSR-7 standard
     * and is provided for Stream's compatibility with pipes.
     *
     * @param int $offset Stream offset
     * @throws RuntimeException on failure.
     */
    public function seekForward($offset)
    {
        // Note that fseek returns 0 on success!
        if (!$this->isSeekableForward() || fseek($this->stream, $offset, SEEK_CUR) === -1) {
            throw new RuntimeException('Could not seek forward in stream');
        }
    }
```

I'd also change isSeekable() implementation - there's no need for isPipe() call, I think, as 'seekable' metadata is guaranteed to be 1 in case of seekable handle and something else (typically 0 but undef for pipes) in other cases.

```
    public function isSeekable()
    {
        if ($this->seekable === null) {
            $this->seekable = false;
            if ($this->isAttached()) {
                $meta = $this->getMetadata();
                $this->seekable = ($meta['seekable'] === 1);
            }
        }

        return $this->seekable;
    }
```

HTH.

**Comment by llvdl** [comment]
Thanks for the feedback, @ddalek.

I replaced `isSeekable()` with your version, but then multiple unit tests fail, so I reverted to the version I submitted before.

**Comment by ghost** [comment]
Argh, I misinterpreted print_r() output. My comparison was unnecessary, the value returned is bool, not int. This will teach me not to rely on print_r(). :)

Still, 'seekable' meta should be false for pipes (despite SEEK_CUR working) and `&& !$this->isPipe()` in isSeekable() adds the cost of fstat() to the first isSeekable() call on verifiably seekable streams. But that's just a tiny perf nitpick (if I'm not talking out of my back again).

**Comment by llvdl** [comment]
For some reason the tests fail on the hhvm machine: `RuntimeException: Could not seek forward in stream` is raised when trying to read from the pipe.

**Comment by llvdl** [comment]
I removed the check for `&& !$this->isPipe()` and the test still pass, so it seems you are right, @ddalek .

**Comment by ghost** [comment]
Cursory search through their issues shows quite a few historical incompatibilities around files and streams, including a new one around fseek() so I'm not terribly surprised. I'll take the time over the weekend to see if there's anything that can and should be done in Slim or if it's HHVM problem.

**Comment by akrabat** [comment]
Tests fail on HHVM.

**Comment by llvdl** [comment]
@I created a virtual machine with hhvm and could reproduce the failing tests. The cause was `isReadable()` returning `false` in hhvm. I added a unit test and a check for `isPipe()` and the tests 

I removed the seekForward implementation for now, as it doesn't seem to work for pipes in PHP. 

**Update:** the reason `isReadable()` returned `false` with hhvm is that the mode is empty when using `stream_get_meta_data()` on a resource opened with `popen()`:

``` php
<?php
$fh = popen('echo 1', 'r');
$meta = stream_get_meta_data($fh);
var_dump($meta['mode']);

stream_get_contents($fh);
pclose($fh);
```

Result using PHP (version 7.0.4): `string(1) "r"`
Result using hhvm (version 3.13.1): `string(0) ""`

Apparently, the mode for the stream cannot be determined using `stream_get_meta_data()` in hhvm in the case of pipes. For what it's worth, the [manual](http://php.net/manual/en/function.stream-get-meta-data.php) does not mention streams using `popen()`, so the behaviour of hhvm is technically not incorrect.

Also, although the metadata provided by hhvm using `stream_get_meta_data()` on a pipe states that the stream is seekable, a `fseek($fh, 1, SEEK_CUR)` seems to always return an error (`-1`) and does not change the stream pointer.

With the current behaviour of hhvm, I think it will be hard to get pipes supported by `Stream` correctly.

**Update**: I have reported the issue at hhvm: https://github.com/facebook/hhvm/issues/7089

File: Slim/Http/Stream.php [hunk 1]
```diff
@@ -73,6 +78,13 @@ class Stream implements StreamInterface
     protected $size;
 
     /**
+     * Is this stream a pipe?
+     *
+     * @var type
+     */
+    protected $pipe;
```
Comment (akrabat) [inline Slim/Http/Stream.php]: Please name this property `isPipe`.

File: Slim/Http/Stream.php [hunk 2]
```diff
@@ -20,6 +20,11 @@
 class Stream implements StreamInterface
 {
     /**
+     * This is octal as per header stat.h
```
Comment (akrabat) [inline Slim/Http/Stream.php:25]: Can you add info on what this particular mode means?

File: Slim/Http/Stream.php [hunk 3]
```diff
@@ -73,6 +78,13 @@ class Stream implements StreamInterface
     protected $size;
 
     /**
+     * Is this stream a pipe?
+     *
+     * @var type
```
Comment (akrabat) [inline Slim/Http/Stream.php]: This is a `boolean` I think

File: Slim/Http/Stream.php [hunk 4]
```diff
@@ -158,6 +170,7 @@ public function detach()
         $this->writable = null;
         $this->seekable = null;
         $this->size = null;
+        $this->pipe = null;
```
Comment (akrabat) [inline Slim/Http/Stream.php]: Should default to `false`.

File: Slim/Http/Stream.php (hunk 1 above)
Comment (llvdl) [inline Slim/Http/Stream.php]: That would collide with the `isPipe` method

File: Slim/Http/Stream.php (hunk 4 above)
Comment (llvdl) [inline Slim/Http/Stream.php]: The value is cached. If `null` the value is determined and then set in `isPipe()`.

File: Slim/Http/Stream.php (hunk 1 above)
Comment (akrabat) [inline Slim/Http/Stream.php]: No. PHP can tell the difference between a property and a method with the same name…

File: Slim/Http/Stream.php (hunk 4 above)
Comment (akrabat) [inline Slim/Http/Stream.php]: Ah. In that case. When is it cleared?

Comment (llvdl) [inline Slim/Http/Stream.php]: In `Stream::detach()`

File: Slim/Http/Stream.php (hunk 1 above)
Comment (llvdl) [inline Slim/Http/Stream.php]: I didn't know that :)

File: Slim/Http/Stream.php (hunk 2 above)
Comment (llvdl) [inline Slim/Http/Stream.php:25]: I expanded the comment

**Pull request merged**